### PR TITLE
Molecule testing

### PR DESCRIPTION
**Pull request by maglub**
This will add molecule tests to our repo.

I have tested with "all" OS:es that we have previously supported, except for CentOS 8, since I could not properly get it to work. It is probably an easy fix, but I rather get this to you right now for review.

I think it will be an easy migration to go from Travis CI to any other runtime if needed.

I am basing my efforts on Jeff Geerling's docker containers with systemd, with the prepare.yml setting up minor things before the tests, that are missing in his container images. One important thing (for me) is to run the tests against the containers as a non `root` user.

### CODE REVIEW

**Comment by maglub** [comment]
Argh, docker hub has rate limits which affects Travis CI (probablay only travis-ci.org, and not travis-ci.com).

**Comment by maglub** [comment]
The Ubuntu 16 build seems to be much slower than the other builds, and is often timing out. I am not sure why, but it might be a memory issue. We are running 5 containers in parallell (postgres 9.6 to 13), so it might be a memory issue. On my Mac, the tests run through within 15 minutes.

File: .travis.yml [hunk 1]
```diff
@@ -1,35 +1,28 @@
...
-language: python
-python: "2.7"
 sudo: required
-dist: trusty
+language: python
+env:
+  - MOLECULE_DISTRO=fedora33   ANSIBLE_ROLES_PATH=.. PY_COLORS='1' ANSIBLE_FORCE_COLOR:=1 DOCKER_TLS_CERTDIR="/certs"
+  - MOLECULE_DISTRO=debian9    ANSIBLE_ROLES_PATH=.. PY_COLORS='1' ANSIBLE_FORCE_COLOR:=1 DOCKER_TLS_CERTDIR="/certs"
+  - MOLECULE_DISTRO=debian10   ANSIBLE_ROLES_PATH=.. PY_COLORS='1' ANSIBLE_FORCE_COLOR:=1 DOCKER_TLS_CERTDIR="/certs"
+#  - MOLECULE_DISTRO=ubuntu1604 ANSIBLE_ROLES_PATH=.. PY_COLORS='1' ANSIBLE_FORCE_COLOR:=1 DOCKER_TLS_CERTDIR="/certs"
```
Comment (otakup0pe) [inline .travis.yml:8]: Do we still intend to support `ubuntu1604` / `centos8` ?

File: .travis.yml [hunk 2]
```diff
@@ -1,35 +1,28 @@
...
-
-  # Idempotence test
-  - ansible-playbook -i tests/docker/hosts -e image_name=${IMAGE_NAME} tests/docker/site.yml | tee idempotence_out
-  - ./tests/idempotence_check.sh idempotence_out
+  - molecule test
+#  - molecule converge
+#  #--- Ubuntu 16 takes quite some time to build, so we tell Travis to run our molecule test wrapped with "travis_wait"
+#  - travis_wait 30 molecule test
 
-notifications:
```
Comment (otakup0pe) [inline .travis.yml:34]: I can set you up on Galaxy for this role I think.

File: molecule/default/molecule.yml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+dependency:
+  name: galaxy
+driver:
+  name: docker
+lint: |
+  set -e
+#  yamllint .
+#  ansible-lint
+platforms:
+#--- postgresql 9.5 no longer supported, https://www.postgresql.org/support/versioning/
```
Comment (otakup0pe) [inline molecule/default/molecule.yml:11]: If we don't intend to support it we can just pull the comment block.

File: tasks/users.yml [hunk 4]
```diff
@@ -13,7 +13,7 @@
     port: "{{postgresql_port}}"
     state: "{{ item.state | default('present') }}"
     login_user: "{{postgresql_admin_user}}"
-  no_log: true
+#  no_log: true
```
Comment (otakup0pe) [inline tasks/users.yml:16]:
```suggestion
  no_log: "{{postgresql_nolog}}"
```
❓

**Review by otakup0pe — COMMENTED** [review]
:cool: :cool:

I guess the next step is moving it to github actions....

File: .travis.yml (hunk 1 above)
Comment (maglub) [reply .travis.yml:8]: Absolutely. I think we can refactor the other pull request for centos8 after we merge this into master.

Comment (maglub) [reply .travis.yml:8]: For ubuntu 16, I am testing different approaches, as it seems we are exhausting the ram on the travis vm and end up in a timeout. Might be something to look into at github actions instead.

File: .travis.yml (hunk 2 above)
Comment (maglub) [reply .travis.yml:34]: Would probably help. I will uncomment the notification, good call.

File: molecule/default/molecule.yml (hunk 3 above)
Comment (maglub) [reply molecule/default/molecule.yml:11]: I would like to use them quite soon, but that is a different work package that requires a different discussion about code guidelines.

File: tasks/users.yml (hunk 4 above)
Comment (maglub) [reply tasks/users.yml:16]: Will fix that. It is a relic from some debugging.

File: molecule/default/molecule.yml (hunk 3 above)
Comment (maglub) [reply molecule/default/molecule.yml:11]: Agree, removing 9.5

**Comment by maglub** [comment]
I think I am "done" for this iteration. I have fixed my code concerning the comments above, found a decent workaround for Ubuntu 16, and the molecule testing feels stable enough for a first release. I will get started with the Github Actions implementation and come back with a new pull request for that.

If we merge this with master, we can ask @chilinot to test his against centos8 using molecule in https://github.com/ANXS/postgresql/pull/509.

@gclough, @otakup0pe: feel free to merge if you are happy with this. Feels better to have you merge it, rather than I pressing the butten myself.

File: molecule/README.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+* fedora33
+* debian9
+* debian10
+* ubuntu1604
+* ubuntu1804
+* ubuntu2004
+
+The images we use are extended with systemd by Jeff Geerling. Before we start the tests, `molecule` runs the `prepare.yml` playbook to:
+
+* Create a user called `ansbile`, with the default group membership of either `wheel` (CentOS, Fedora), or `sudo` (Debian, ubuntu)
```
Comment (UnderGreen) [inline molecule/README.md:32]: `ansbile` -> `ansible`

File: molecule/README.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
+# Introduction
+
+This directory is the home of the test playbooks:
+
+* ../tests/prepare.yml => setting up a VM or container with the minimum requirements (that usually is there, but not always in a container)
+* ../tests/playbook.yml => running a test against a VM or container (default postgres version: 13)
+
+# Molecule
+
+The default tested version is postgresql 9.6, 10, 11, 12, and 13 on Ubuntu 20.04. We are currently _not_ linting anything.
```
Comment (UnderGreen) [inline molecule/README.md:10]: We are currently _not_ linting anything. -> Linting is disabled for the tests.

File: molecule/README.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+```
+lint: |
+  set -e
+#  yamllint .
+#  ansible-lint
+```
+
+You can override this with setting the environment variable MOLECULE_DISTRO to one of:
```
Comment (UnderGreen) [inline molecule/README.md:19]: I am not sure I followed what exactly one can override here?

File: molecule/README.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+* centos7
+* centos8
+* fedora33
+* debian9
+* debian10
+* ubuntu1604
+* ubuntu1804
+* ubuntu2004
+
+The images we use are extended with systemd by Jeff Geerling. Before we start the tests, `molecule` runs the `prepare.yml` playbook to:
```
Comment (UnderGreen) [inline molecule/README.md:30]: I would provide here a link to JG repo with images Dockerfiles. Or, at least docker hub link?

Comment (UnderGreen) [reply molecule/README.md:30]: Before we start the tests, `molecule` runs the `prepare.yml` playbook to: -> Prior to the testing, `molecule` runs the `prepare.yml` playbook to

File: molecule/README.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+* Create a user called `ansbile`, with the default group membership of either `wheel` (CentOS, Fedora), or `sudo` (Debian, ubuntu)
+* Install a couple of packages that Jeff Geerling did not install in his container images, that we need in order to test the role properly
```
Comment (UnderGreen) [inline molecule/README.md:33]: that we need -> that are needed

File: molecule/README.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+```
+
+# Examples
+
+Examples:
+
+```
+molecule create
+molecule converge
+molecule test
```
Comment (UnderGreen) [inline molecule/README.md:82]: I believe `molecule test` provides whole cycle from create to destroy.
```
test         Test (dependency, lint, cleanup, destroy, syntax, create, prepare, converge, idempotence, side_effect, verify, cleanup, destroy).
```

File: tests/README.md [hunk 11]
```diff
@@ -0,0 +1,8 @@
+# Introduction
+
+For more docu about molecule testing, see ../molecule/README.md
```
Comment (UnderGreen) [inline tests/README.md:3]: docu -> documentation

File: .travis.yml [hunk 12]
```diff
@@ -1,35 +1,25 @@
...
+  - MOLECULE_DISTRO=fedora33   ANSIBLE_ROLES_PATH=.. PY_COLORS='1' ANSIBLE_FORCE_COLOR:=1 DOCKER_TLS_CERTDIR="/certs"
+  - MOLECULE_DISTRO=debian9    ANSIBLE_ROLES_PATH=.. PY_COLORS='1' ANSIBLE_FORCE_COLOR:=1 DOCKER_TLS_CERTDIR="/certs"
+  - MOLECULE_DISTRO=debian10   ANSIBLE_ROLES_PATH=.. PY_COLORS='1' ANSIBLE_FORCE_COLOR:=1 DOCKER_TLS_CERTDIR="/certs"
+  - MOLECULE_DISTRO=ubuntu1604 ANSIBLE_ROLES_PATH=.. PY_COLORS='1' ANSIBLE_FORCE_COLOR:=1 DOCKER_TLS_CERTDIR="/certs"
+  - MOLECULE_DISTRO=ubuntu1804 ANSIBLE_ROLES_PATH=.. PY_COLORS='1' ANSIBLE_FORCE_COLOR:=1 DOCKER_TLS_CERTDIR="/certs"
+  - MOLECULE_DISTRO=ubuntu2004 ANSIBLE_ROLES_PATH=.. PY_COLORS='1' ANSIBLE_FORCE_COLOR:=1 DOCKER_TLS_CERTDIR="/certs"
+  - MOLECULE_DISTRO=centos7    ANSIBLE_ROLES_PATH=.. PY_COLORS='1' ANSIBLE_FORCE_COLOR:=1 DOCKER_TLS_CERTDIR="/certs"
+#  - MOLECULE_DISTRO=centos8    ANSIBLE_ROLES_PATH=.. PY_COLORS='1' ANSIBLE_FORCE_COLOR:=1 DOCKER_TLS_CERTDIR="/certs"
+before_install:
+  - echo 'DOCKER_OPTS="$DOCKER_OPTS --debug --registry-mirror=https://mirror.gcr.io"' | sudo tee -a /etc/default/docker
```
[the remark this answers is not in the document]
Comment (maglub) [reply .travis.yml:14]: Yes, that is the intention, but I am not completely sure it fix the problem.

File: .travis.yml [hunk 13]
```diff
@@ -1,35 +1,25 @@
 ---
-
-language: python
-python: "2.7"
 sudo: required
-dist: trusty
+language: python
+env:
```
[the remark this answers is not in the document]
Comment (maglub) [reply .travis.yml:4]: True, but I could not come up with any other way.

File: .travis.yml (hunk 2 above)
Comment (otakup0pe) [reply .travis.yml:34]: Hmm still mucking with galaxy. I think you can only do org level admin.

**Comment by maglub** [comment]
Closing this as a duplicate of #511

**Pull request closed without merging**